### ISSUE TITLE
Support localized fields in API responses

### ISSUE THREAD

**Issue by willdurand** [mozilla/addons-frontend#6422]
In order to fully switch to API v5, we need to support the new localized fields. Instead of directly getting a value for any localized field, we will always receive an object with at least one key/value (the key is the locale, the value is the localized content).

The API `v5` should be used to see those localized fields.

See also: https://github.com/mozilla/addons-server/issues/8794

**Comment by eviljeff** [comment]
this will unblock #3049 when implemented

**Comment by jvillalobos** [comment]
This would be nice to have for accessibility.

**Comment by eviljeff** [comment]
Updated issue to reflect localized fields are in api/v5 now.

**Comment by bobsilverberg** [comment]
@willdurand I'm going to work on this as a code quality issue. Do you think we should assume that we're only getting one locale back from an API request (as we always pass a `lang` as a query parameter), or is the idea here that we should assume we could be getting multiple locales returned, and we should use the current `lang` to look up the value in the returned object? If the latter, what would we do if we don't find a match for the current lang? I know that should never happen, but if we're coding to deal with multiple returned langs, we should also code to deal with that scenario.

**Comment by bobsilverberg** [comment]
Hmm, I see in code manager that we first try to get the string for the current lang, and if that fails then we return the string for the first lang we find. I guess that logic would make sense here too.

**Comment by eviljeff** [comment]
Fwiw the API will only ever return a single translation if the `lang` param is sent.

**Comment by bobsilverberg** [comment]
After having spent some time on this, working on code that would try to find the correct string for the current `lang`, it seems like that's just not going to be worth it. It's not complicated, but it does involve passing the `lang` around in a lot of places where we currently don't need it, so it will result in a lot of code changes. Because we always pass a `lang` to API requests, it doesn't seem necessary to support the ability to look up a `lang` in the returned object. We can still deal with a situation in which we get more than one `lang` back by always just using the first one.

I think this is a much more practical approach, but I just want to double check with @willdurand, who opened this issue, to make sure it's satisfactory.

**Comment by willdurand** [comment]
I was going to say "let's do this" but then your proposal made me realize that the API response for localized fields is less than ideal. I filed https://github.com/mozilla/addons-server/issues/16069 to change it.

**Comment by bobsilverberg** [comment]
> I was going to say "let's do this" but then your proposal made me realize that the API response for localized fields is less than ideal. I filed [mozilla/addons-server#16069](https://github.com/mozilla/addons-server/issues/16069) to change it.

I like your suggestion in https://github.com/mozilla/addons-server/issues/16069, but I don't think that really affects this issue much, except for the fact that we'd need to change the type definitions. I think we'll still want to assume we're getting back a single language and just use the first one, as I suggested above. Do you agree with that @willdurand?

**Comment by willdurand** [comment]
> I like your suggestion in [mozilla/addons-server#16069](https://github.com/mozilla/addons-server/issues/16069), but I don't think that really affects this issue much, except for the fact that we'd need to change the type definitions. I think we'll still want to assume we're getting back a single language and just use the first one, as I suggested above. Do you agree with that @willdurand?

That won't be possible because if a localized string isn't available for a request locale, then we will have 3 properties in the object (the request locale, the default value and the fallback property). So we cannot really "use the first one".

**Comment by bobsilverberg** [comment]
> > I like your suggestion in [mozilla/addons-server#16069](https://github.com/mozilla/addons-server/issues/16069), but I don't think that really affects this issue much, except for the fact that we'd need to change the type definitions. I think we'll still want to assume we're getting back a single language and just use the first one, as I suggested above. Do you agree with that @willdurand?
> 
> That won't be possible because if a localized string isn't available for a request locale, then we will have 3 properties in the object (the request locale, the default value and the fallback property). So we cannot really "use the first one".

That's not the way I read the change. Even with a fallback, the object will look like:
```json
{ "field": { "en-US": "value", "fr": null, "fallback": "en-US" } }
```

As long as the actual value being returned is the first one, can't we just take the first property, which in this case would be the `en-US` value? If the server did find a match for our requested `lang`, I assume the object would look like:
```json
{ "field": { "fr": "value" } }
```

so again, taking the first property would work. WDYT?

**Comment by eviljeff** [comment]
Yeah, we could have the first item in the dict be the default, which would also make it backwards compatible with what code-manager assumes.  (Though it would reduce any benefit mozilla/addons-server#16069 would bring once the primary consumer - addons-frontend - just assumes the first item anyway)

**Comment by willdurand** [comment]
mmm and so we'd base our approach on order of properties in a JSON object converted into a JavaScript object?

**Comment by willdurand** [comment]
> Yeah, we could have the first item in the dict be the default, which would also make it backwards compatible with what code-manager assumes. (Though it would reduce any benefit [mozilla/addons-server#16069](https://github.com/mozilla/addons-server/issues/16069) would bring once the primary consumer - addons-frontend - just assumes the first item anyway)

I think code-manager won't be affected, we only support one locale, which is `en-US`.

**Comment by bobsilverberg** [comment]
> mmm and so we'd base our approach on the order of properties in a JSON object converted into a JavaScript object?

Yes we would, but only because we know we are always going to get a single language back. It would then be a requirement on the server side to return the value for that fallback language as the first property in the object, which shouldn't be difficult to satisfy. If we don't do that, then we're back to having addons-frontend try to look up the requested `lang` in the returned object, which means passing the `lang` around *a lot*, which complicates our code, even though we know we're only ever getting a single language back.

I think the issue here is the fact that we're requesting a change to the API which we don't actually need, from addons-frontend's perspective. I agree it makes sense in terms of the API design, which is why I said I like the suggestion, but the bottom line is that addons-frontend will never use it. Not only do we always send the `lang` param, and therefore always expect a single language returned, but we also don't care about which language is returned to us. We won't do anything differently if what we're getting back is in the fallback language.

**Comment by bobsilverberg** [comment]
> > Yeah, we could have the first item in the dict be the default, which would also make it backwards compatible with what code-manager assumes. (Though it would reduce any benefit [mozilla/addons-server#16069](https://github.com/mozilla/addons-server/issues/16069) would bring once the primary consumer - addons-frontend - just assumes the first item anyway)
> 
> I think code-manager won't be affected, we only support one locale, which is `en-US`.

I believe that @eviljeff is correct. Code manager won't be affected, as long as we return the real value in the first property. Code manager does try to find a translation for the current lang, and if one is not found, returns the first one found. If we want Code manager to behave "properly" with respect to the API change, we should probably update its code, after https://github.com/mozilla/addons-server/issues/16069 lands, to try to use a fallback lang. This would just be a simple change to a utility function so it's probably worth doing.

**Comment by willdurand** [comment]
> Yes we would, but only because we know we are always going to get a single language back. It would then be a requirement on the server side to return the value for that fallback language as the first property in the object, which shouldn't be difficult to satisfy. If we don't do that, then we're back to having addons-frontend try to look up the requested `lang` in the returned object, which means passing the `lang` around _a lot_, which complicates our code, even though we know we're only ever getting a single language back.

mmm although I am sure order is likely guaranteed, this is very fragile. We should not make such assumptions.

> I think the issue here is the fact that we're requesting a change to the API which we don't actually need, from addons-frontend's perspective. I agree it makes sense in terms of the API design, which is why I said I like the suggestion, but the bottom line is that addons-frontend will never use it. Not only do we always send the lang param, and therefore always expect a single language returned, but we also don't care about which language is returned to us. We won't do anything differently if what we're getting back is in the fallback language.

API v5 is experimental so I don't see why we'd need to rush and implement something that does not work for us. This issue is about supporting the new localized fields in API v5 but it's clearly not ready so I think we should wait until v5 gets fixed. There is no need to upgrade to the current API v5.

> Not only do we always send the lang param, and therefore always expect a single language returned, but we also don't care about which language is returned to us. We won't do anything differently if what we're getting back is in the fallback language.

Not yet but that's why API v5 introduces this change, because for a11y, we'll have to indicate when the content isn't for the requested locale.

**Comment by bobsilverberg** [comment]
It's true there is no urgency for this, so I guess we can put it back on the back burner for now. It does sound like we need some more issues open for it, as this one didn't really call out the need to indicate when content isn't in the current locale. If we want to support the possibility of multiple locales being returned (and looking up the correct one), and also knowing when a fallback language was used, I think it will result in bigger changes, so those requirements should be documented somewhere. This issue was pretty sparse in terms of describing what's needed.

**Comment by eviljeff** [comment]
If we need some more issues open, can someone create them to detail exactly what changes are needed?  So far we just have https://github.com/mozilla/addons-server/issues/16069 - if that's all that's required, great, but it's not clear from this issue if that's enough on the api side.  

While it's true there's no project deadline adding urgency, this is the single thing that is preventing addons-frontend from moving to v5 (which in turn blocks v4 from replacing v3 as the stable api version, which then blocks us from deprecating v3) so there's a bunch of old code we need to maintain and work around.  It's been blocked for more than 2 years already (mozilla/addons-server#9467 and others that moved the feature gate from v4, to v4dev, to v5, each time it was put off)

**Comment by willdurand** [comment]
See also: https://github.com/mozilla/addons-frontend/issues/3049

**Comment by willdurand** [comment]
I added the `needs: api` label because this issue depends on fixing the API first, see: https://github.com/mozilla/addons-server/issues/16069.

The main reason behind the API change in v5 is captured in #3049 but it is an old issue. That being said, even if we don't fix #3049, we still need to do some work in addons-frontend to support the "new" localized fields, mainly because adopting v5 in addons-frontend would allow us to mark v4 as stable and eventually deprecate v3.

I see two options to move forward:

1. we revert the localized field change in v5, meaning it is the same behavior in v4, v5, etc. It's not ideal because the change _does_ make sense from an API design perspective.
2. we fix the API and then we update addons-frontend to use the new localized fields. This will require some work but it does not seem that big. I would strongly prefer to consume the API correctly rather than relying on a hack (described in https://github.com/mozilla/addons-frontend/issues/6422#issuecomment-736543977)

#3049 is a separate issue: if we go for (1), we wontfix it; if we go for (2), we might be able to work on it eventually (that would be feasible from a technical perspective).

**Comment by eviljeff** [comment]
I don't advocate it, but if we go with (1) we're effectively deciding never to use the localized field change in v5, so we should remove the support from addons-server.  The actual code isn't that complicated, but most tests assume that v5 is the default and the API response will be a dict so changing those will be a non-trivial amount of work.

**Comment by bobsilverberg** [comment]
I don't advocate 1 either. I am fine with 2, and also fine with making changes to "properly" consume the API. I just wanted to be clear that, although it's not complicated, it will involve a lot of code changes.

**Comment by AlexandraMoga** [comment]
After testing the frontend functionality as a whole, with more focus on localizations, there were no regressions found. 

I'm marking this issue as verified fixed on stage.
